src/python/pants/backend/codegen/protobuf/java/java_protobuf_library.py

```diff
@@ -18,6 +19,9 @@
 class JavaProtobufLibrary(ImportJarsMixin, JvmTarget):
```

Comment:
Hm... so, I think the fact that this implements `ImportJarsMixin` might just be legacy... I think that the `remote_sources` target type could be used with a `java_protobuf_library` target without this.

We have a tiny internal example (https://gist.github.com/stuhood/779666f75a6909440857acf9f2365910) of a macro that uses remote sources to compile thrift that is extracted from a jar, which implies that integration with the target type isn't necessary.

If it makes things easier things easier to stop implementing `ImportJarsMixin` here, would suggest that.

Reply:
I spent some time trying to make a wrapper which would preserve the current behavior using a context-aware object factory, since I think it should be possible but I'm not super familiar with `_type_aliases`/`target_types` as in the very clear example you've shown (my next step is going to be to get more familiar with that so I can apply it directly, this is mostly a note to self).

I got to the point of:
```python

class MaybeRemoteJavaProtobufLibrary(object):
  """The `remote_sources` type assumes things about the registrar, so quack like a registrar."""

  def __init__(self, parse_context):
    self._parse_context = parse_context

  def __call__(name, imports=None, **kwargs):
    """Create a "remote" protobuf library, which requires two target definitions.
    With the exception of the `imports` parameter, all arguments match `java_protobuf_library`.
    :param list imports: A list of specs pointing to `jar_library()` targets containing protobuf.
    """
    if not imports:
      self._parse_context.create_object(JavaProtobufLibrary, name=name, **kwargs)
    else:
      unpacked_name = '{}-unpacked-jar-{}'.format(name, stable_json_hash(imports))
      self._parse_context.create_object('unpacked_jars',
                                        name=unpacked_name,
                                        libraries=imports,
                                        include_patterns=['**/*.proto'])
      self._parse_context.create_object('remote_sources',
                                        name=name,
                                        dest=JavaProtobufLibrary,
                                        sources_target=':{}'.format(unpacked_name),
                                        args=kwargs)
```

which I registered in `test_java_protobuf_library.py` as:
```python
  @classmethod
  def alias_groups(cls):
    return BuildFileAliases(
        context_aware_object_factories={
          'java_protobuf_library': MaybeRemoteJavaProtobufLibrary,
        },
        targets={
          'jar_library': JarLibrary},
        objects={
          'jar': JarDependency,
          'scala_jar': ScalaJarDependency,
        }
      )
```

but I then got this while running `./pants test tests/python/pants_test/backend/codegen/protobuf/java -- -vs`:
```
self = <pants_test.backend.codegen.protobuf.java.test_java_protobuf_library.JavaProtobufLibraryTest testMethod=test_wrong_import_type1>
                     
                         def test_wrong_import_type1(self):
                           self.add_to_build_file('BUILD', dedent('''
                               java_protobuf_library(name='foo',
                                 sources=[],
                                 imports=[':not_jar']
                               )
                         
                               java_protobuf_library(name='not_jar',
                                 sources=[],
                               )
                               '''))
                     >     target = self.target('//:foo')
                     
                     .pants.d/pyprep/sources/cc4d1bd31e29e8b06f94704640f389b790bca529/pants_test/backend/codegen/protobuf/java/test_java_protobuf_library.py:73: 
                     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
                     .pants.d/pyprep/sources/cc4d1bd31e29e8b06f94704640f389b790bca529/pants_test/test_base.py:532: in target
                         self.build_graph.inject_address_closure(address)
                     .pants.d/pyprep/sources/cc4d1bd31e29e8b06f94704640f389b790bca529/pants/engine/legacy/graph.py:205: in inject_address_closure
                         self.inject_addresses_closure([address])
                     .pants.d/pyprep/sources/cc4d1bd31e29e8b06f94704640f389b790bca529/pants/engine/legacy/graph.py:213: in inject_addresses_closure
                         for _ in self._inject_specs(specs):
                     .pants.d/pyprep/sources/cc4d1bd31e29e8b06f94704640f389b790bca529/pants/engine/legacy/graph.py:270: in _inject_specs
                         subjects)
                     /opt/twitter_mde/package/eepython27/e4cdee382b83e4185a42b54120b1b11021cec1d8285cc9023ef2dabf8c9af24d/lib/python2.7/contextlib.py:35: in __exit__
                         self.gen.throw(type, value, traceback)
                     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
                     
                     self = <pants.engine.legacy.graph.LegacyBuildGraph object at 0x10c8eaed0>
                     
                         @contextmanager
                         def _resolve_context(self):
                           try:
                             yield
                           except Exception as e:
                             raise AddressLookupError(
                     >         'Build graph construction failed: {} {}'.format(type(e).__name__, str(e))
                             )
                     E       AddressLookupError: Build graph construction failed: ExecutionError 1 Exception encountered:
                     E       Computing Select(Specs(dependencies<=tuple>=(SingleAddress(directory=u'', name=u'foo'),), matcher<=SpecsMatcher>=SpecsMatcher(tags<=tuple>=(), exclude_patterns<=tuple>=())), =TransitiveHydratedTargets)
                     E         Computing Task(transitive_hydrated_targets, Specs(dependencies<=tuple>=(SingleAddress(directory=u'', name=u'foo'),), matcher<=SpecsMatcher>=SpecsMatcher(tags<=tuple>=(), exclude_patterns<=tuple>=())), =TransitiveHydratedTargets, true)
                     E           Computing Task(addresses_from_address_families, Specs(dependencies<=tuple>=(SingleAddress(directory=u'', name=u'foo'),), matcher<=SpecsMatcher>=SpecsMatcher(tags<=tuple>=(), exclude_patterns<=tuple>=())), =BuildFileAddresses, true)
                     E             Computing Task(parse_address_family, Dir(path=), =AddressFamily, true)
                     E               Throw(Failed to parse BUILD:
                     E       __call__() got multiple values for keyword argument 'name')
                     E                 Traceback (most recent call last):
                     E                   File "/Users/dmcclanahan/tools/pants/.pants.d/pyprep/sources/cc4d1bd31e29e8b06f94704640f389b790bca529/pants/engine/native.py", line 409, in extern_generator_send
                     E                     res = c.from_value(func[0]).send(c.from_value(arg[0]))
                     E                   File "/Users/dmcclanahan/tools/pants/.pants.d/pyprep/sources/cc4d1bd31e29e8b06f94704640f389b790bca529/pants/engine/build_files.py", line 61, in parse_address_family
                     E                     address_mapper.parser))
                     E                   File "/Users/dmcclanahan/tools/pants/.pants.d/pyprep/sources/cc4d1bd31e29e8b06f94704640f389b790bca529/pants/engine/mapper.py", line 55, in parse
                     E                     raise MappingError('Failed to parse {}:\n{}'.format(filepath, e))
                     E                 MappingError: Failed to parse BUILD:
                     E                 __call__() got multiple values for keyword argument 'name'
```

I'm going to come back to this tomorrow.

Reply:
A middle ground might be to continue to extend `ImportJarsMixin`, but to deprecate that extension (IE, put in a conditional deprecation in `java_protobuf_library` that looks for a deprecated arg) with a recommendation to use `remote_sources`?

Reply:
Done in 25e1546, and made the rest into #7111!